exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java

```diff
@@ -215,6 +206,7 @@ public BatchHolder() {
           MaterializedField outputField = materializedValueFields[i];
           // Create a type-specific ValueVector for this value
           vector = TypeHelper.getNewVector(outputField, allocator);
+          int columnSize = new RecordBatchSizer.ColumnSize(vector).estSize;
```

Comment:
I wonder what is wrong if estSize is 0 when there is no data. If there is no data for a column, why would we want to add it's value width to outgoing row width ?

Reply:
Agree that the "sizer" should return the size of a fixed-width column even if there are 0 rows. Thanks for catching and fixing this. Though, I will point out that if you get a 0-length batch, just throw it away and read another; no useful purpose is served by doing anything with an empty batch (except to pass it along for schema purposes if this is the first batch: the so-called "fast schema" that is supposed to work, but seems to not be fully implemented...)

I've found that Drill behaves poorly and erratically when trying to allocate zero-sized vectors. Best not to do that. If your batch size is 0, don't allocate vectors would be my general advice to avoid lots of grief.

Reply:
@ppadma I thought estSize represented the estimated column width. For FixedWidth vectors we know the exact column width, so why can't we use the exact value? Also why are there two different things for measuring column sizes, when do you use RecordBatchSizer and when do you use TypeHelper?

Reply:
@ilooner That is the point. If we know the exact value, why do we need RecordBatchSizer ? we should use RecordBatchSizer when we need to get sizing information for a batch (in most cases, incoming batch). In this case, you are allocating memory for value vectors for the batch you are building. For fixed width columns, you can get the column width size for each type you are allocating memory for using TypeHelper.getSize. For variable width columns, TypeHelper.getSize assumes it is 50 bytes.  If you want to adjust memory you are allocating for variable width columns for outgoing batch based on incoming batch, that's when you use RecordBatchSizer on actual incoming batch to figure out the average size of that column.  You can also use RecordBatchSizer on incoming batch if you want to figure out how many values you want to allocate memory for in the outgoing batch. Note that, with your change, for just created value vectors, variable width columns will return estSize of 1, which is not what you want.

Reply:
I can think of three reasons to use the sizer:

* Type logic is complex: we have multiple sets of rules depending on the data type. Best to encapsulate the logic in a single place. So, either 1) use the "sizer", or 2) move the logic from the "sizer" to a common utility.
* Column size is tricky as it depends on `DataMode`. The size or a `Required INT` is 4. The (total memory) size of an `Optional INT` is 5. For a `Repeated INT`? You need to know the average array cardinality, which the "sizer" provides (by analyzing an input batch.)
* As discussed, variable-width columns (`VARCHAR`, `VARBINARY` for HBase) have no known size. We really have to completely forget about that awful "50" estimate. We can only estimate size from input, which is, again, what the "sizer" does.

Of course, all the above only works I you actually sample the input.

A current limitation (and good enhancement) is that the Sizer is aware of just one batch. The sort (the first user of the "sizer") needed only aggregate row size, so it just kept track of the widest row ever seen. If you need detailed column information, you may want another layer: one that aggregates information across batches. (For arrays and variable-width columns, you can take the weighted average or the maximum depending on your needs.)

Remember, if the purpose of this number is to estimate memory use, then you have to add a 33% (average) allowance for internal fragmentation. (Each vector is, on average, 75% full.)

Reply:
Thanks for the info @paul-rogers . Since we want to have the RecordBatchSizer be a one stop shop for size related information I propose to do the following to address @ppadma 's concerns:

 - Remove changes to estSize in the RecordBatchSizer that I introduced.
 - Add a field called knownSize to the RecordBatchSizer. This field will contain the known column width for FixedWidthVectors and will be -1 for things like varchars and repeated columns. Then we can add a method getKnownSize() for a column size. This method will return the known size if available and will throw an exception if the column is not FixedWidth. So this method should only be used by the caller when they know they are dealing with FixedWidth columns as is the case with HashAgg for now (things will probably change when strings will be stored in varchars instead of objects). The rest of RecordBatchSizer will remain as is.
 - Add a constant called FRAGMENTATION_FACTOR (1.33) that people can use to multiply their custom crafted batch sizes by when they are estimating batch sizes from scratch. In HashAgg I would multiply the custom batch size I compute by this factor.

Thoughts?